Sources/Plasma/Apps/plClient/Mac-Cocoa/main.mm

```diff
@@ -457,20 +449,27 @@ - (void)startClient
 {
     PF_CONSOLE_INITIALIZE(Audio)
 
-    self.plsView.delegate = self;
+    self.renderLayer = gClient->GetPipeline()->GetRenderLayer();
+    _renderLayer.backgroundColor = NSColor.blackColor.CGColor;
+
+    self.window.contentView.layer = _renderLayer;
+    self.window.contentView.wantsLayer = YES;
     // Create a window:
 
     // Window controller
     [self.window setContentSize:NSMakeSize(800, 600)];
     [self.window center];
     [self.window makeKeyAndOrderFront:self];
-    self.renderLayer = self.window.contentView.layer;
-    
-    [self.renderLayer addObserver:self
-                       forKeyPath:@"device"
-                          options:NSKeyValueObservingOptionNew | NSKeyValueObservingOptionInitial
-                          context:DeviceDidChangeContext];
-    
+
+#ifdef PLASMA_PIPELINE_METAL
+    if ([_renderLayer respondsToSelector:@selector(device)]) {
+        [_renderLayer addObserver:self
+                           forKeyPath:@"device"
+                              options:NSKeyValueObservingOptionNew | NSKeyValueObservingOptionInitial
+                              context:DeviceDidChangeContext];
+    }
+#endif
```

Comment:
This is a good change. I think that this probably does extra checking. It could simply check the layer type and then check for device - in since CAMetalLayer already has device defined. The respondsToSelector would be unnecessary (wouldn't want some other device function floating in unintentionally in a different layer type.) The ifdef might be redundant in since it's a dynamic check, but we could skip the dynamic check if Metal wasn't present.

Longer term - we need a common code path for GL and Metal to handle screen and device changes. I think screen right now is all CGDisplay based so that's agnostic to the underlying API.

Reply:
As far as I can tell, the only thing this is currently being used for it to print the Metal device name in the window title for debugging purposes, so I'm hopeful it can be dropped entirely at some point in the future.

Reply:
Strangely - this has been useful. On Intel MacBook Pros I've run into situations in which Plasma is running on the wrong GPU, so it's a nice sanity check. However it could be maybe re-implemented against the enumeration code that already exists in the engine and may already be storing the same data.

There's another reason this is a helpful debug hint. Plasma may need to switch GPUs mid run (which can be supported with Plasma's GPU crash recovery mechanisms.) So double checking the GPU device would be helpful for proofing that out. But I haven't gotten eGPU ejection or Plasma-follows-your-current-displays-GPU yet.